Create an Ingest API

Bargs opened the pull request:
**This feature was removed from 5.0. It does not currently have a target release because it will most likely require some changes.** 

Closes #5199
## Documentation

If you like Postman, import [this collection](https://www.getpostman.com/collections/0b9fffdc7bf629d786e6) with example requests to play around with.
### Endpoints
### POST /api/kibana/ingest

Create a new index pattern and index template. Template mappings are given sensible defaults based on their type. Fields of type `number` will be created as a `double` in elasticsearch. Strings will be mapped as multi fields with an analyzed version and a non-analyzed version (called `{fieldName}.raw`). Fields with a `.` in their name will be mapped as object type fields with mappings included for each of their subfields.

_Note:_ You can only create a template if your pattern does not match existing indices.
#### Payload Schema

Each payload is a JSON object.

The object has the following properties:
- `id`: String (Required) This will be the ID of the index pattern that gets created.
- `title`: String (Required)
- `time_field_name`: String (optional, only needed for time based index patterns)
- `interval_name`: String (optional)
- `fields`: Array (required)
- `field_format_map`: Object (optional, map of non-default formatters to user for fields)
- `not_expandable`: Boolean (optional)

An object in the `fields` array must have the following attributes:
- `name`: String (required)
- `type`: String (required)

Valid values for `type` include:
- string
- date
- boolean
- number
- geo_point
- geo_shape
- ip
##### Errors
- 400 Bad Request for an invalid payload
- 409 Conflict when a pattern or template with the given ID already exists
- 409 Conflict when the pattern matches existing indices
##### Example

Request:

```
POST /api/kibana/ingest

{
          "id": "logstash-*",
          "title": "logstash-*",
          "time_field_name": "@timestamp",
          "fields": [{
            "name": "ip",
            "type": "ip"
          }, {
            "name": "@timestamp",
            "type": "date"
          }, {
            "name": "agent",
            "type": "string"
          }, {
            "name": "bytes",
            "type": "number"
          },
          {
            "name": "geo.coordinates",
            "type": "geo_point"
          },
          {
            "name": "geo.src",
            "type": "string"
          },
          {
            "name": "geo.dest",
            "type": "string"
          },
          {
            "name": "geo.srcdest",
            "type": "string"
          }]
}
```

Response:

```
Status 204 
```
#### DELETE /api/kibana/ingest/{id}

Deletes the index pattern with the given id and its associated index template.
##### Errors
- 404 Not Found if the given pattern does not exist
##### Example

Request:

```
DELETE /api/kibana/ingest/logstash-*
```

Response:

```
Status 200

{"success": true}
```
### Done
- Moved API endpoints into a plugin.
- Implemented basic initial version of all the endpoints.
- Error handling 
  Need to handle errors that the elasticsearch client might throw at me
  https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/errors.html
- Create an elasticsearch index template when an index pattern is created
- Ignore type specific mappings, use [default](https://www.elastic.co/guide/en/elasticsearch/reference/master/default-mapping.html)
  - Accept json payload that ES expects for mappings. 
  - If a pattern matches no existing indices, a template will be created.
  - If a pattern matches existing indices, a template will not be created.
  - If a user is trying to update an existing pattern with a matching template, we should only accept new field mappings and reject changes to existing ones (add a warning in the UI that uses this API to describe this behavior?) (going to punt on this until we feel like we actually need it because it's going to be rather complicated).
- Payload validation
  Create a Joi schema
- Testing
  Primarily going to rely on functional tests for API testing, do unit testing where appropriate. 
- Security
  Should accept API tokens for auth once #5267 is merged.
  Until then, how should we secure this? It's a bit different than most ways we access elasticsearch right now because it accesses the .kibana index, but it will be initiated from the frontend. Should look at how the current pattern creation screen does it with a shielded cluster.

kjbekkelund commented:
What are pros and cons of plugin vs core? If there are no big cons to putting this in a plugin, that seems like the best way to go (smaller core is nice).

Bargs commented:
@kjbekkelund yup I agree. The last couple days I've been learning more about Hapi and the current state of the plugin system and that's the direction I'm going to move in. Should have some updated code today.

Bargs commented:
@rashidkpc what level of validation do you think we should do on the request body JSON, if any? For instance should we parse the "fields" field and make sure it looks like it's formatted correctly? 

From a security standpoint, it looks like the /elasticsearch endpoint allows the user to send pretty much anything to the .kibana index. This new API forces the .kibana index and index-pattern type for all requests since it only deals with index patterns, but can you think of any potentially malicious payloads we should try to detect and reject?

I also have some questions related to ES index template creation in the description of this PR that I'd appreciate if you could take a look at.

rashidkpc commented:
I can't think of anything particularly malicious you could add to a mapping. 

Fields is stored as a string to keep elasticsearch from parsing it into real elasticsearch fields, causing the cluster state to be much larger than it would need to be. This is probably not a good move, we probably could've just turned off indexing of that field in the mapping. Not sure if that works on object level fields.

Bargs commented:
@rashidkpc we'd really only be creating templates for new index patterns, right? Could we deprecate and hide the `number` type and only allow new index patterns to use one of the elasticsearch types?

rashidkpc commented:
We certainly could, though we'd need to update a lot of code that looks for 'number' specifically.

Bargs commented:
@rashidkpc I think the API is pretty much functionally complete. I still have a few things I'd like to do (testing, investigate security/shield integration, refactor some common code into modules), but I'd love it if you could take a look when you get a chance next week and provide any feedback you might have. Thanks!

kjbekkelund commented on src/plugins/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,221 @@
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
+      const mappings = _.omit(_.mapValues(_.indexBy(req.payload.fields, 'name'), (value) => {
+        return value.mapping;
+      }), _.isUndefined);
```
For readability I usually prefer something like this when I use a lot of lodash methods:

```javascript
_(req.payload.fields)
    .indexBy('name')
    .mapValues(value => value.mapping)
    .omit(_.isUndefined)
    .value();
```

Bargs replied:
Didn't know about that syntax. Mind blown!

spalger commented on src/plugins/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,221 @@
+    path: '/api/index-patterns',
+    method: 'GET',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
```
You probably want to use `callWithRequest` here, so 401s are properly forwarded to the user. Right @simianhacker?

Bargs replied:
@spalger Is that all I need to do to make this work with Shield?

Bargs replied:
I believe I've eradicated all direct use of the client now.

spalger commented on src/plugins/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,221 @@
+        }),
+        getMappings(pattern, client),
+        stitchPatternAndMappings
+      ).then(removeDeprecatedFieldProps)
```
I think this function could use a makeover. The `.then()`s are miss-indented, and I think that `Promise.all().spread()` would read better than `.join()`.

spalger commented on src/plugins/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,221 @@
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
```
I'm confused by the regex here, I don't see `moment()` anywhere, or the `toIndexList()` method... does this really support index patterns which rely on those types of index patterns? Elasticsearch doesn't understand the `logstash-[YYYY.MM.DD]` format.

Bargs replied:
That line needs to be removed from the PUT handler, it got copy pasted from the POST handler. In the POST handler the `isWildcard` declaration is only intended to tell me if this pattern will match more than 1 index. If so, we create an index template. If there will only ever be one index we don't bother. As far as I understand the * and [] syntaxes are the only way an index pattern will match multiple indices so that's what I tested for, I'm open to suggestions if there's a better way to accomplish this.

spalger replied:
Well the `[]` syntax is actually a moment-specific syntax, and therefore will never match anything if sent to elasticsearch. I think it make sense to have the consumer of the API decide if the template should be made by requiring an additional parameter, perhaps:

```
# no index template
POST /api/indexPatterns/logstash-*

# create index template with logstash-* wildcard
POST /api/indexPatterns/logstash-*?indexTemplateWildcard=logstash-*
```

This means that by default the API will only do the minimum, create an index pattern, but you can opt-into index template creation easily. It also means that the front-end, which has the weird moment-template-as-index-pattern, can convert that pattern into the proper es-compatible wildcard.

Bargs replied:
I see what you're saying now, the index template won't match anything if [] is used because ES doesn't understand it. I'm hesitant to give the user the ability to set the template pattern themselves because it would allow the template and index pattern to get out of sync. As I understand it the pattern and template should match all the same indices. Given that the `[]` syntax is deprecated anyways I'd lean towards only creating a template when `*` is used. @rashidkpc any thoughts on this?

spalger replied:
Well, yes, this would allow the index patterns to get out of sync but that can sometimes be desired (as is the case with the `[]` patterns).

I personally feel that creating an index pattern shouldn't automatically create an index template, It doesn't feel very REST-y to have that kind of side effect by default. I do like that the API will create the index template for you should you want it.

Bargs replied:
That was my initial reaction too, but the way I'm implementing it I think it is REST-y. The schema for the index pattern REST resource that I'm defining includes the (optional) mappings for the fields  (one item on my to-do list is to add a Joi schema and beef up the documentation around this in the ticket). Those mappings are stored in an index template, not in the .kibana index and that's ok. Different parts of a REST resource can be persisted in different places (e.g. in two separate tables in a SQL database). Creating the template isn't a side effect since the mappings are a part of the index pattern REST resource by definition. If a client POSTs an index pattern without mappings, no template will be created, so they have to intentionally opt-in to this behavior. 

That said, the idea to create an index template was Rashid's, so that's why I was trying to get his input :smile:

spalger replied:
Well I guess I can address @rashidkpc then.

The fields are actually stored in the .kibana index [by this code](https://github.com/Bargs/kibana/blob/indexPatternApi/src/plugins/api/index_pattern_endpoints.js#L105-L107), so the index tempate being created is not necessary or actually connected to the index pattern in any way. Will deleting an index pattern delete any matching index templates by default?

I am primarily opposed to this behavior because I don't think anyone would expect "POST /api/indexPattern" to create an index template. It clearly translates to "create an index pattern". "POST /api/indexPattern?withIndexTemplate=logstash-*" on the other hand translates clearly.

rashidkpc replied:
This is a tough call. In a perfect world we wouldn't even need our own `indexPattern` we'd could just use the template as the single source of truth. The tough bit that no matter what we're going to end up with 3 sources of possible mappings for field *foo*:

1. The indexPattern contains the mapping for *foo* as kibana sees it. 
2. The elasticsearch mapping contains fields that already exist in indices that already exist. 
3. The template. The field as it *might* be in indices that don't exist yet. This is crucial for new data coming in, to make sure fields get mapped correctly.

That said, I'm with @spalger here, I don't see any harm in being explicit.

Bargs replied:
The core question here is whether POSTing a pattern should also create a template. I don't have a strong opinion there, but if we do decide to separate them I would prefer to expose a /api/template endpoint and allow the client to use that instead of tacking on a parameter to the index-pattern endpoint.

Let me try to explain the current code a bit more though. The client already has to opt-in to template creation. They do this by providing mappings for the fields in their pattern. You'll notice the code you linked to omits mappings, that's because they're stored in the template, not in .kibana. The documentation for this endpoint should say "Create an index pattern. Optionally provide mappings for your fields and we'll create an index template for you as well." A field in the payload might look like this: 

```
referrer: {
  count: 1,
  scripted: false,
  mappings: {
    type: string,
    doc_values: false,
    index: analyzed
  }
}
```

The schema for the index pattern returned from this endpoint is a bit different than the current index patterns. Client code that uses this API would need to support this new format. I did this intentionally so that the mappings from the template and the info from .kibana could be glued into a single cohesive REST resource. The definition of an index pattern in terms of this API is (old index pattern + mappings). But again, that's just my attempt to make things REST-y if we want to create a template when POSTing to /index-patterns.

spalger commented on src/plugins/api/package.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "api",
```
I think `api` might be a bit generic 😉

spalger replied:
We have a convention of using `/api/${plugin.name}` for a plugins routes. I feel like we should pick a better name for this plugin and then follow the naming convention everyone else does.

spalger replied:
What about "index-patterns"? We could then move the index pattern stuff out of `ui` and into this plugin.

Bargs replied:
So my thinking here was that index-patterns are a core part of Kibana (seeing as how kibana can't operate without one, and they're stored in the .kibana index). This is different than a plugin like Marvel, which stores its data in its own indices. I called the plugin `api` because I imagined it would be the location for future Kibana APIs (e.g. for visualizations or dashboards). Creating a separate plugin for each Kibana REST resource seems like overkill to me. Perhaps these endpoints should actually go in the kibana plugin? Then we could have endpoints like:

`/kibana/api/index-patterns`
`/kibana/api/visualizations`
`/kibana/api/dashboards`

Those match the marvel endpoints nicely e.g. `/marvel/api/v1/clusters`

spalger replied:
If the intention is to tie these api's to kibana then I think they should go into the kibana plugin.

Also, the convention is `/api/{plugin}` not `/{plugin}/api`, does the marvel team have it backwards?

Bargs replied:
Agreed, I'll move it into the kibana plugin.

And yes, Marvel currently exposes api endpoints as `/marvel/api`

spalger replied:
Marvel updated their routes to use the `api/plugin` convention

Bargs replied:
API moved into the kibana plugin and names updated to follow conventions.

spalger commented on src/plugins/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,221 @@
+  });
+
+  server.route({
+    path: '/api/index-patterns',
```
What do you think about sticking with `camelCase` in the urls?

Bargs replied:
I'm really not a fan of camelCase in urls. I think hyphens and underscores are much more prevalent in api design and since this is an interface we're exposing to the outside world, I don't think our internal naming conventions should dictate how we form these urls.

spalger replied:
Well we currently only expose `/api/status` so you're kind of setting the standard. I'm going to open an issue for voting.

rashidkpc replied:
i'm strongly against camelCase in HTML, I much prefer dashes. We already sluggify objects with dashes, let's just stick to them

Bargs replied:
Yeah I'm cool with snake_case, I voted for that on the ticket as well.

Bargs replied:
Now updated to snake case.

rashidkpc commented:
@Bargs can you add a thorough description of how the API is used to the pull description. I know you have an initial description in #5199, but I'd like to see the final implementation here.

Bargs commented:
@rashidkpc Updated. I think I covered all of the functionality as it currently stands. Let me know if you have any questions or concerns. My goal was to have the API combine all the info from the .kibana index, index templates, and indices in a way that's mostly seamless to the client, without duplicating data in any of those three locations.

Bargs commented:
@rashidkpc @spalger Now that the tests are complete and passing I've slapped the official review label on this guy. The only outstanding items from the last batch of feedback are:

1. Is using callWithRequest all I need to do to work properly with Shield?
2. Resolution to the discussion about template creation.

I updated the PR description with some thorough documentation as well so please give that a read as you review.

rashidkpc commented:
@Bargs can you update the pull description to fit the new routes?

rashidkpc commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+      return error;
+    }
+  };
+
```
Can you break these routes out into their own files, under an index_pattern directory? Then just include them in this file?

rashidkpc commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+
+export default function (server) {
+
+  let handleESError = function (error) {
```
Also break this out, and write mocha tests for it.

rashidkpc replied:
Also this function seems useful for a lot more than the index pattern API, probably belongs in `kibana/server/lib`

spalger commented on src/plugins/kibana/server/lib/schemas/index_pattern_schema.js:
```diff
@@ -0,0 +1,31 @@
+module.exports = {
+  post: Joi.object({
+    title: Joi.string().required(),
+    timeFieldName: Joi.string(),
```
keep in mind that these need to be snake_cased

Bargs replied:
:snake: ified

rashidkpc commented on src/plugins/kibana/server/lib/stitch_pattern_and_mappings.js:
```diff
@@ -0,0 +1,12 @@
+const _ = require('lodash');
+
+module.exports = function stitchPatternAndMappings(pattern, mappings) {
```
Can you describe what this should do and why. Unit tests with examples would help here.

spalger commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+            return result._source;
+          }),
+        getMappings(pattern, req),
+        stitchPatternAndMappings)
```
I had to count parenthesis to find that the `join()` call ends at this line. I think this section should be formatted like so:

```js
Promise.join(
  callWithRequest(req, 'get', params)
  .then((result) => {
    result._source.fields = JSON.parse(result._source.fields);
    return result._source;
  }),
  getMappings(pattern, req),
  stitchPatternAndMappings
)
.then(removeDeprecatedFieldProps)
.then(
  function (pattern) {
    reply(_.isArray(pattern) ? pattern[0] : pattern);
  },
  function (error) {
    reply(handleESError(error));
  }
);
```

Bargs replied:
@spalger I remember you had a comment recommending an alternative to join, but I can't find that comment anymore. Seems like it got deleted somehow. What was your recommended alternative?

spalger commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*');
+      const mappings = _(req.payload.fields)
+        .indexBy('name')
```
please align these lines with the `const` on line 118, [ref](https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md#chaining-operations)

spalger commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+      };
+
+      callWithRequest(req, 'create', patternCreateParams)
+        .then((patternResponse) => {
```
All of the `.then()` calls need to be indented with the line the operate on.

spalger commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+          if (!isWildcard || _.isEmpty(mappings)) {
+            return patternResponse;
+          }
+          else {
```
Please convert this to use an early return rather than wrapping everything in an else branch.

spalger commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+                if (matchingIndices) {
+                  throw Boom.conflict('Cannot create an index template if existing indices already match index pattern');
+                }
+                else {
```
another unnecessary else branch

spalger commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+    path: '/api/kibana/index_patterns/{id}',
+    method: 'PUT',
+    handler: function (req, reply) {
+      if (_.isEmpty(req.payload)) { return reply(Boom.badRequest('Payload required')); }
```
for single line if statements please omit the `{` brackets `}`

spalger commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+
+      if (!_.isEmpty(mappings)) {
+        reply(Boom.badRequest('Mappings cannot be updated'));
+      } else {
```
another place I would use early return

spalger commented on src/plugins/kibana/server/lib/get_mappings.js:
```diff
@@ -0,0 +1,39 @@
+const Promise = require('bluebird');
+const {templateToPattern, patternToTemplate} = require('./convert_pattern_and_template_name');
+
+module.exports = function getMappings(pattern, req) {
```
Methods that require a request object can generally be refactored to accept one or two more parameters (like a bound version of `callWithRequest` in this instance) and in the process become simpler to test.

spalger commented on src/plugins/kibana/server/routes/api/index_pattern_endpoints.js:
```diff
@@ -0,0 +1,258 @@
+    method: 'POST',
+    handler: function (req, reply) {
+      if (_.isEmpty(req.payload)) { return reply(Boom.badRequest('Payload required')); }
+      const validation = Joi.validate(req.payload, indexPatternSchema.post);
```
The hapi [route options](http://hapijs.com/api/8.8.1#route-options) accept validation configuration that is perfect for this type of work.

rashidkpc commented:
`GET /api/index_patterns`
Retrieve the list of all index patterns. 
Mapping info is retrieved from the matching index template if it exists. If it does not exist, the mappings are retrieved from the indices themselves.

Why are we retrieving and showing the mapping? Wouldn't we would want the kibana index pattern here, possibly with the template?

`POST /api/index_patterns`
Create a new index pattern.

Feels like this should take `POST /api/index_patterns/{id}` no? 

Also I tried to use this API and I get a 500 when I try to send the example resource. Am I using this incorrectly? Also what happens if the index pattern already exists? Overwritten? What if the pattern doesn't exist but a mapping for the index does exist, is it handled the way creating a pattern from the UI currently is?

```
TypeError: Uncaught error: Cannot assign to read only property 'fields' of {
  "title": "usagov",
  "timeFieldName": "@timestamp",
  "fields": [
    {
      "name": "geo.coordinates",
      "mapping": {
        "type": "geo_point",
        "index": "not_analyzed",
        "doc_values":false
      }
    }
  ]
}
```

rashidkpc commented:
Can you fill in the description of the pull with more complete examples of how the APIs are used and the expected responses?

Bargs commented:
@rashidkpc 

> Why are we retrieving and showing the mapping? Wouldn't we would want the kibana index pattern here, possibly with the template?

Maybe that wasn't worded well. `GET index_patterns` will pull back all of the index patterns from .kibana in the format described in the PR. The index pattern will include the mapping for each field, pulled from the template if it exists, from the index otherwise. My thinking was that we shouldn't duplicate info like type, doc_values, indexed status in the .kibana index pattern if we're creating templates, but I needed to fall back on the info in the index if the template didn't exist.

> Feels like this should take POST /api/index_patterns/{id} no?

That's a good question... I was using the index pattern title as the ID since that's how it works today. However, based on the discussion in this PR #5428 I'm thinking it might be useful to allow an ID that's different than the title. If we don't allow the ID and title to be different, it seems redundant to require the client to put it in both the body and the URL. Thoughts?

> Also I tried to use this API and I get a 500 when I try to send the example resource. Am I using this incorrectly? 

Hmm, I was able to POST your example successfully. Did you post to `/api/kibana/index_patterns` (just updated the route in the description, sorry about that)? Is there anything about your .kibana mapping that would make `fields` readonly? Also if you're hitting the API with Postman or curl you'll need to start the server with `--server.xsrf.disableProtection=true` for now, but that should give you a different error.

> Also what happens if the index pattern already exists? Overwritten? What if the pattern doesn't exist but a mapping for the index does exist, is it handled the way creating a pattern from the UI currently is?

If the pattern already exists you get a 409 conflict. What happens when the index or index template already exist depends on whether or not you include mappings in your request body. 

*No mappings* - Index pattern gets created. When you do a GET for this pattern the mappings will be retrieved from the existing template or index.

*Mapping included* - 409 conflict with a message describing the problem. This is basically what happens any time you try to send mappings to the API but they already exist in some form. Since updating mappings is so problematic, I think we should only accept them when creating a fresh new pattern/template with no existing matching indices.

rashidkpc commented:
> Hmm, I was able to POST your example successfully. Did you post to `/api/kibana/index_patterns` (just updated the route in the description, sorry about that)? Is there anything about your .kibana mapping that would make `fields` readonly? Also if you're hitting the API with Postman or curl you'll need to start the server with `--server.xsrf.disableProtection=true` for now, but that should give you a different error.

Possibly that the index already exists? Lets chat tomorrow

Bargs commented:
@rashidkpc I beefed up the PR description with more examples, info on error codes, and a downloadable Postman collection. Let me know if there's anything else helpful that you think I could add.

Bargs commented:
@rashidkpc I noticed during the meeting that the ES guys create and update the actual asciidoc as a part of their pull. Do you think it would be beneficial if I did the same with the documentation I've been adding to the PR description, that way it doesn't get buried after this PR is closed?

Bargs commented:
@rashidkpc @spalger I was talking to Rashid last week and he mentioned that the main reason we currently store field mapping info inside the index pattern in .kibana is to avoid having to request potentially thousands of field mappings from the indices. Currently I have the new GET endpoints grabbing the mapping info directly from the indices if no template exists, so I've been thinking about this problem. Since we already require the user to manually refresh the index pattern, I'm thinking it would be cleaner to set the appropriate cache-control headers, and potentially cache the result server side. It looks fairly easy to cache responses with Hapi, do we do this at all currently? Are there any reasons you think I shouldn't take this approach?

The alternative would be to store the mappings inside .kibana upon pattern creation and provide some sort of refresh mechanism like we do today, but now that we have a backend server in the stack it feels kinda hackey to implement our own caching in this way.

Bargs commented:
I believe I've incorporated all of the feedback now. The only outstanding question I have is the one about caching above.

Bargs commented:
Well the browser cache idea is out. I thought I could manually request a fresh version by setting `cache-control: no-cache` in the request when the user clicked the 'refresh fields' button. It works in Chrome, but firefox has a 7 year old bug that I guess they aren't going to fix :( https://bugzilla.mozilla.org/show_bug.cgi?id=443098

I still think server side caching with Hapi is preferable to storing a second copy of the mappings in .kibana. The normalized index patterns are fairly small so it shouldn't take much memory to cache them, and we can simply invalidate the cache any time the pattern is updated via the API.

Bargs commented:
I created a separate pull with a rough implementation of what I'm thinking for caching the index patterns: #5575

trekr5 commented:
Hi,

When I tried to use the endpoint 
    GET /api/index_patterns

I get the following error:
    {
      "error" : "ElasticsearchIllegalArgumentException[No feature for name 9index_patterns]]",
      "status" : 400
}

I'm running Elasticsearch 1.7.3

Can you tell me how I can gain access to these endpoints?

Bargs commented:
Thinking about this more, one upside to "caching" the normalized mapping info in the .kibana index pattern (like it is today) is that we wouldn't have to refactor the existing frontend index pattern code right now. As I look at that code more, it seems like a non-trivial task to switch it over to the new API. What do you think @rashidkpc ? Is a refactor of the existing frontend code in scope here, or in the interest of time should we only use the API for the new ingest screens?

rashidkpc commented:
@Bargs thats my biggest concern right now. Its not that i don't want to refactor a lot of the front end, its that I'm concerned about timeline and increasing scope of an already pretty complex project. 

I think it would be useful to have refactoring the front end as part of a phase 2.

Bargs commented:
@rashidkpc I agree, it's probably too much to fit into phase 1. Also the more I think about it, we should probably hold off on refactoring until we've established some best practices for consuming these new APIs. I created a discussion ticket (#5596) to talk about that subject. The new ingest pages can serve as an opportunity to experiment a bit. Once we get to phase 2 we should have a better idea of how to clean up the old code.

So on that note, I'm going to update the API to store the index-patterns in .kibana in a backwards compatible manner.

rashidkpc commented:
Sounds good!

Bargs commented:
I've finished updating the API to store the index patterns in a backwards compatible manner. I also improved the payload schemas so that they now treat the template as a separate but related document. Next week I need to update the documentation and create some additional tests for the new functionality and then we should finally be able to get this thing across the finish line.

Bargs commented:
Alright @spalger and @rashidkpc, I think this thing is just about at the finish line. I know you're both busy this week, but if you could take another look when you get a chance I'd appreciate it.

Some changes to be aware of:
* The index patterns created by the API should now be stored in a way that's backwards compatible with the current Kibana frontend code. 
* I've made significant changes to the API document schema which make the pattern's relationship to its index template more explicit. The `mappings` property I added to the pattern fields is gone now, instead a client will pass mappings as a part of a related template resource.
* The new API document schema also prepares us for decoupling the index pattern's title from its ID.

The PR description and Postman examples should be up to date with all the changes, but let me know if you have any questions.

Bargs commented:
@spalger and I chatted today and we came up with a few updates that I'll implement:

* Store the template_id attribute in camelCase in ES, only snake case it in the API payload
* In GET index_patterns, combine all template requests into one by concatenating the template ids into a single comma separated string
* Use the index pattern ID instead of the title for pattern matching
* Should DELETE throw an error if template deletion is requested, but the specified index pattern has no related template? The current code works this way, but Spalger and I are leaning towards making it more lenient.

spalger commented:
Just sat down with @Bargs and dug through all the nitty gritty. Looks good to me!

**Just a thought**: Should we merge this into a feature branch rather than master? My thinking is that we might not want to commit to shipping this specific API until after we have actually used it. Ideally this can get merged into a feature branch and then work can start on the node injest UI. Once it is determined that the indexPattern API is solid then the feature branch can be merged into master.

simianhacker commented:
@Bars @spalger I agree... we need to get this merged and start using it. A
feature branch makes perfect sense.

On Wed, Dec 16, 2015 at 5:24 PM Spencer <notifications@github.com> wrote:

> Just sat down with @Bargs <https://github.com/Bargs> and dug through all
> the nitty gritty. Looks good to me!
>
> *Just a thought*: Should we merge this into a feature branch rather than
> master? My thinking is that we might not want to commit to shipping this
> specific API until after we have actually used it. Ideally this can get
> merged into a feature branch and then work can start on the node injest UI.
> Once it is determined that the indexPattern API is solid then the feature
> branch can be merged into master.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/kibana/pull/5213#issuecomment-165298412>.
>

Bargs commented:
@spalger I implemented the changes we talked about (listed in my previous comment). If you could give it one final look I'd be much obliged.

@simianhacker I think this PR is 99.9% there. I've implemented all the changes Spencer and I talked about today, once he gives the LGTM I'd like to pass it to @rashidkpc for one final look, and then when we meet on Friday I'll quiz him about merging this into master vs. a feature branch.

spalger commented:
LGTM

Bargs commented:
@rashidkpc @spalger @epixa I'm summarizing the items we talked about here so I don't forget next week. Let me know if I've missed anything or if anything is wrong.

- [ ] remove concept of template resource in the API payload
- [ ] create template if fields array exists, based only on the field mapping info Kibana cares about and stores in the index pattern today (indexed status, doc_values, type, name).
- [ ] throw error if fields array is missing
- [ ] throw error if matching indices already exist
- [ ] Make sure it works with shield
- [ ] Ensure created patterns are backwards compatible

Some questions I have after giving our discussion some more thought:

* Are `indexed`, `analyzed`, and `doc_values` required properties for each field? If not do we need to set defaults or are they optional in ES index template mappings as well?
* What was the field @epixa added to index patterns? Is it required or optional?
* Should we allow updates to fields? Would it also update the template? What happens if the user adds a field to their ES docs and clicks Refresh Fields on the existing index pattern edit page, so the field exists in the pattern but not the template? I'm tempted to remove the PUT endpoint completely since it's complicated and it's not strictly necessary for the ingest UI. Maybe it's something we worry about in phase 2 when we refactor the existing frontend index pattern code.
* Besides field mappings, are there any other settings we should set in the new index template by default?
* Should DELETE remove both the pattern and template? I'm thinking yes.

epixa commented:
The field is the boolean `notExpandable`, and it is optional.

Bargs commented:
@rashidkpc @spalger I've run into an issue with generating a template from the index pattern. There's no way to tell the difference between a nested object and a multi-field. For instance, if we POST an index pattern with a field called `machine.os`, there's no way to know if the user wants 

```
"machine": {
  "properties": {
    "os": {
...
```

or 

```
"machine": {
  "type": "string",
  "fields": {
    "os": {
      "type": "string"
...
```

In the context of the ingest UI, I guess it's safe to always assume it's a nested object since we're dealing with raw data. But when a client does a GET, they'll get an index pattern with multi-fields included (e.g. `machine.os.raw`), and POSTing that index pattern to a fresh kibana would give them a totally different pattern/template if we assume those are nested fields. That's really not good. I'm having a hard time thinking of any good solutions at the moment, do you guys have any ideas?

spalger commented:
I would like to reiterate my opinion from our most recent review of this pr:

I feel like we need two separate APIs

 1. `/api/kibana/index_pattern`
   - mostly just CRUD for index-pattern objects
   - it could automatically discover the field list if one isn't provided (maybe v2?)
 2. `/api/kibana/ingest_config`
   - accepts a config object containing whatever is useful
     - sample data
     - field names
     - data sources to pull
     - etc.
   - creation of this config could look something like:
     1. create an index template in elasticsearch
     1. ? trigger the index template by indexing some sample data ?
     1. create an index pattern pointing to the newly created indices
     1. populate the field list of that index pattern based on the discovered fields in the newly created indices
    - deletion of a config could cleanup anything it created (index template, index pattern) if desired
    - these documents could be considered immutable, updates are not allowed (only deletes)

Taking this approach makes @Bargs problem a lot simpler, the `ingest_config` object would simply need a way to communicate what is and isn't a multi-value field. Since the `ingest_config` object isn't trying to disguise itself as an index pattern it can be formatted to meet the needs of the ingest UI.

This also makes the `index_pattern` API something we can easily update the existing UI to use, since it doesn't have to bend to meet the needs of the ingest UI.

Bargs commented:
Thanks for the feedback @spalger

A couple of thoughts:

1. I like the idea of the index_pattern API being a simple crud interface that doesn't cause unexpected side-effects like the creation of an index template. That's what led me to my previous design, where the template was an explicitly related resource that the client could include or not at their discretion.

2. The main critique of the previous design (correct me if I'm wrong) was that it required the client to understand too much about ES. An ingest_config API does provide an abstraction but I'm afraid it will be a leaky one. Today we need a way to communicate the concept of a multi field. Tomorrow we might need aliases, the next day dynamic templates, and so on. We'll end up re-inventing the entire index template syntax and then a client will need an intimate knowledge of both ES templates and Kibana's special dialect for them. That's another reason why I ended up making templates a first class citizen in the previous design, because I couldn't think of a useful abstraction that would give clients all the power they need, without simply layering complexity on top of the existing ES template API.

I think this API has uncovered a lot of unanswered questions about what it means to be a Kibana API in general. Here are a few questions, which I thought I had answers to at various times, but now I'm not so sure:

* Who's the intended client for the API? Kibana frontend, third parties, both? This should have a clear answer, not "sort of both" which is where I started and I think led to a lot of this churn.
* Should the client need to know *anything* about ES? If so, how much? I've been fighting with some cognitive dissonance on this one, where on one hand we say Kibana APIs should hide ES details, but some of the features we’re building are straight up GUIs for ES features that will need very granular access to ES APIs.
* How strictly do we want to follow REST conventions? Sometimes the coupling with ES will make it difficult to create clean, independent resources.
* What use cases does the API need to support? If we take the position that Kibana has total control over ES and you're not allowed to muck about with the ES APIs directly, that leads to a simpler design but conflicts with the majority of today's use cases. Even if we wanted to head in that direction, we would still need to worry about backwards compatibility for any APIs being built today.

@rashidkpc I'm happy to continue exploring these questions, but I'm getting seriously worried about the front end portion of the project. I think the best course of action at this point would be to mark this API as internal only and cut out everything that isn't necessary for the ingest UI. In other words, everything except for a single POST endpoint that creates a pattern and a simple template based on that pattern. Everything else is probably better suited for a phase 2 that could coincide with a refactor of the existing front end index pattern code. If you agree, I should be able to have that set of functionality ready for the review tomorrow.

spalger commented:
> The main critique of the previous design (correct me if I'm wrong) was that it required the client to understand too much about ES.

Agreed, I feel like we should remove all elasticsearch-specific syntax/configuration/object definitions from the front-end code and expose structured, documented, and tested APIs that contain all of that complexity. This will make our UI simpler (and maybe easier to make portable) as well as provide an abstraction point for us to implement improvements/backwards-compat solutions that make sense for Kibana but not for Elasticsearch.

> Today we need a way to communicate the concept of a multi field. Tomorrow we might need aliases, the next day dynamic templates, and so on.

Agreed, but that is my intention. We need a place to implement each of those things but it's likely that we will want to rename certain things, add smart defaults based on certain criteria, etc, and I feel like an api like `ingest_config` is the place to do that. Alternatively we implement those translations at the UI layer and they have to be reimplemented by all consumers of the API (even if we aren't planning for more than one consumer, that's not ideal).

> That's another reason why I ended up making templates a first class citizen in the previous design, because I couldn't think of a useful abstraction that would give clients all the power they need

I can't remember saying this before, which is my fault, but I see the UI basically being serialized to a JSON object and then sent to the API. This means that whatever abstraction makes sense in the UI is what our API would expose. This might be a horrible idea, but it's truly what I imagine the API being, essentially a series of AJAX form posting endpoints...

---

We just had a face-to-face after I wrote that, and we talked a lot, so I'm going to go ahead and leave this here so I we can summarize.

Bargs commented:
Summary of what we talked about on the call, please comment if anyone feels I missed something:

* I'll reduce this API to only include the feature set required by the ingest UI.
* This API is intended for internal use only.
* This API should not be seen as an example of best practices for future Kibana APIs.
* Everyone agrees that ES specifics should be abstracted away as much as possible, but there's still a lot of debate about how much we can abstract and what those abstractions should look like.
* While we'd all like to set a good precedent for future APIs, we can't allow perfection to get in the way of completion of features.

Bargs commented:
@rashidkpc @spalger 

Just pushed all the updates we discussed on the call. I also updated the PR title/description and Postman examples, so everything should be totally up to date except for the shield permissions. I'll get with someone to sort the Shield stuff out as soon as I can, in the meantime please give the code one more look!

spalger commented on src/plugins/kibana/server/lib/create_mappings_from_pattern_fields.js:
```diff
@@ -0,0 +1,38 @@
+      };
+    }
+
+    _.set(mappings, field.name.replace('.', '.properties.'), mapping);
```
replace only replaces the first instance of the pattern. I recommend using splice+join (which turns out to be more efficient as well as concise.

```js
field.name.split('.').join('.properties.')
```

Bargs replied:
Doh. Updated.

spalger commented on src/plugins/kibana/server/routes/api/ingest/register_post.js:
```diff
@@ -0,0 +1,101 @@
+          .then(() => {
+            throw templateError;
+          }, () => {
+            throw new Error(`index-pattern ${indexPatternId} created successfully but index template
```
This message should include the error message of the original error.

Bargs replied:
The template creation error message or the pattern deletion error message?

spalger replied:
Both I suppose

Bargs replied:
Done.

spalger commented on src/plugins/kibana/server/routes/api/ingest/register_post.js:
```diff
@@ -0,0 +1,101 @@
+        return callWithRequest(req, 'indices.exists', {index: indexPatternId})
+        .then((matchingIndices) => {
+          if (matchingIndices) {
+            throw Boom.conflict('Cannot create an index pattern via this API if existing indices already match the pattern');
```
Why not do this check first?

Bargs replied:
Great catch. This was left over from when the API allowed patterns to be created for existing indices, but not templates. Updated it to do the check before creating anything.

spalger commented on tasks/config/run.js:
```diff
@@ -24,6 +24,23 @@ module.exports = function (grunt) {
+        quiet: false,
+        failOnError: false
+      },
+      cmd: /^win/.test(platform) ? '.\\bin\\kibana.bat' : './bin/kibana',
```
You can just use the `binScript

Bargs replied:
Updated this for both tasks that weren't using the variable.

spalger commented on tasks/config/intern.js:
```diff
@@ -8,6 +8,12 @@ module.exports = function (grunt) {
     },
-    dev: {}
+    dev: {},
+    api: {
```
We talked about this previously, but I really don't think we need intern/an entire browser for these tests. If you want to do that in a separate PR I'm fine with it.

Bargs replied:
Yeah I'll probably refactor these when I have some more time. They don't start a browser since the intern classifies them as unit tests, but in hindsight they definitely don't need to rely on the intern at all.

spalger commented on src/plugins/kibana/server/lib/case_conversion.js:
```diff
@@ -0,0 +1,15 @@
+const _ = require('lodash');
+
+module.exports = {
+  convertToSnakeCase: function (object) {
```
`convertToSnakeCase` isn't used anywhere

Bargs replied:
Probably will be useful as soon as someone adds and API with a GET endpoint though. I don't see any harm in leaving it in, do you?

spalger commented on src/plugins/kibana/server/lib/case_conversion.js:
```diff
@@ -0,0 +1,15 @@
+    });
+  },
+
+  convertToCamelCase: function (object) {
```
I'm tempted to say that in order for this to be it's own helper it should be deep. I would just put it into the single file that uses it and call it something like `shallowKeysToCamelCase()`

Bargs replied:
It was originally shallow because I didn't want it modifying ES specific names in the mappings. That's not an issue now, but I figured once someone needed it they would add a `convertToSnakeCaseDeep` method to this module.

spalger replied:
Works for me, would you mind mentioning keys and or shallow in the function name?

epixa replied:
If we want to follow the path of libraries like lodash, helpers should be shallow by default and be suffixed with `Deep` in the event that they are recursive.

spalger replied:
In some cases that's the case, `_.merge` is an exception, and in general documentation never hurts.

Bargs replied:
Originally my intent was to follow lodash's convention (shallow by default), but it's true that lodash is inconsistent. `_.isEqual` is also deep by default. That inconsistency and the confusion it just caused here is probably reason enough to be more verbose in our own method names, so I updated them to `keysToSnakeCaseShallow` and `keysToCamelCaseShallow`.

spalger commented on src/plugins/kibana/server/lib/schemas/resources/index_pattern_schema.js:
```diff
@@ -0,0 +1,23 @@
+  time_field_name: Joi.string(),
+  interval_name: Joi.string(),
+  not_expandable: Joi.boolean(),
+  fields: Joi.array().items(
```
fields needs to have at least one field it would seem https://github.com/hapijs/joi/blob/master/API.md#arrayminlimit

Bargs replied:
Updated the schema and added an API test for it as well.

spalger commented on src/plugins/kibana/server/lib/__tests__/create_mappings_from_pattern_fields.js:
```diff
@@ -0,0 +1,81 @@
+
+    _.forEach(function (mapping) {
+      if (mapping.type !== 'string') {
+        expect(_.isEqual(mapping, {
```
I don't think this does what you think it does. You probably have to do:

```js
expect(mapping).is.eql({
  ...
});
```

Bargs replied:
I'm using lodash's isEqual because I want to test for deep and strict equality. Unfortunately expect.js's is.eql() only checks for loose equality, and there's no assertion that's both recursive and strict.

spalger commented on src/plugins/kibana/server/lib/__tests__/create_mappings_from_pattern_fields.js:
```diff
@@ -0,0 +1,81 @@
+
+    _.forEach(function (mapping) {
+      if (mapping.type === 'string') {
+        expect(mapping.to.have.property('fields'));
```
calling expect with a boolean does not actually make an assertion. Try

```js
expect(mapping).to.have.property('fields');
```

Bargs replied:
Typo :(

spalger commented on src/plugins/kibana/server/lib/__tests__/create_mappings_from_pattern_fields.js:
```diff
@@ -0,0 +1,81 @@
+  it('should set the same default mapping for all non-strings', function () {
+    let mappings = createMappingsFromPatternFields(testFields);
+
+    _.forEach(function (mapping) {
```
you're not iterating anything :)

spalger commented on src/plugins/kibana/server/lib/__tests__/create_mappings_from_pattern_fields.js:
```diff
@@ -0,0 +1,81 @@
+    expect(mappings).to.have.property('geo');
+    expect(mappings.geo).to.have.property('properties');
+    expect(mappings.geo.properties).to.have.property('coordinates');
+    expect(_.isEqual(mappings.geo.properties.coordinates, {type: 'geo_point'}));
```
bad assertion

spalger commented on src/plugins/kibana/server/lib/__tests__/create_mappings_from_pattern_fields.js:
```diff
@@ -0,0 +1,81 @@
+  it('should give strings a multi-field mapping', function () {
+    let mappings = createMappingsFromPatternFields(testFields);
+
+    _.forEach(function (mapping) {
```
bad forEach

spalger commented on src/plugins/kibana/server/routes/api/ingest/register_delete.js:
```diff
@@ -0,0 +1,33 @@
+        callWithRequest(req, 'delete', deletePatternParams),
+        callWithRequest(req, 'indices.deleteTemplate', {name: patternToTemplate(req.params.id)})
+            .catch((error) => {
+              if (!error.status || error.status !== 404) {
```
You can set `ignore: [404]` when you call the client to prevent 404 responses from being considered errors, might clean this up a bit.

spalger commented on src/plugins/kibana/server/routes/api/ingest/register_delete.js:
```diff
@@ -0,0 +1,33 @@
+                throw error;
+              }
+            })
+      ])
```
indentation

Bargs commented:
@spalger Just pushed some updates, I think I covered all of your feedback, please take another look!

spalger commented:
LGTM

Bargs commented:
@spalger Whoooooo! @rashidkpc one final look for you as well?

rashidkpc commented:
Implementation looks good, can you squash the commits?

kmoe commented:
Is the Postman collection up to date? When I run the GET request against my local Kibana 5, it complains about a missing `kbn-version` header, and still gives a `400` when I add one in.

Bargs commented:
@kmoe the Postman examples should be up to date, it was working for another team member before. However it sounds like you're getting an old version, because there's no longer a GET endpoint. You might try deleting the postman collection and re-importing it, I've seen the sync act pretty flaky in the past.

megastef commented:
How to use this API? Why is a kbnVersion Header required? And why is ther a browser version check for an API? 
Here are the things I've tried:
```
 curl -XPOST localhost:5601/api/kibana/ingest -d @pattern.txt
{"statusCode":400,"error":"Bad Request","message":"Missing kbn-version header"}MacBook-Pro-2:kibana stefan$ 
MacBook-Pro-2:kibana stefan$ curl -XPOST -H "kbn-version: 4.4.0" localhost:5601/api/kibana/ingest -d @pattern.txt
{"statusCode":400,"error":"Bad Request","message":"Browser client is out of date, please refresh the page"} 
curl -XPOST -H "kbn-version: 4.4.0" -H "mime-type: application/json" -A "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:43.0) Gecko/20100101 Firefox/43.0"  localhost:5601/api/kibana/ingest -d @pattern.txt
{"statusCode":400,"error":"Bad Request","message":"Browser client is out of date, please refresh the page"} 
```

epixa commented:
@megastef This change only exists in master, which is version `5.0.0-snapshot` rather than `4.4.0`.

megastef commented:
I wonder more about the error messages ""Browser client is out of date, please refresh the page" - there was no complain about the version I used in the header (only when not present). And the API does not allow to specify the Kibana-Index (which is configured in the kibana.yml - static/gloabal config), while an API should give the freedom to be used for more general things.

epixa commented:
The "browser out of date" error you're getting actually *is* the error message from the version mismatch.  When the error message was created, there were no APIs to access, so you'd only ever see it if you were using Kibana from your browser.  We should update the error message.

epixa commented:
I created issue #6021 for updating the error message.

Bargs commented:
Also note that this API is primarily meant for Kibana's own internal use
right now. It's possible to hit it from outside of Kibana's front end code,
but you will run into some rough edges like this.
On Wed, Jan 27, 2016 at 10:08 AM Court Ewing <notifications@github.com>
wrote:

> I created issue #6021 <https://github.com/elastic/kibana/issues/6021> for
> updating the error message.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/kibana/pull/5213#issuecomment-175674810>.
>

tamsky commented:
> The "browser out of date" error you're getting actually is the error message from the version mismatch.

For us to solve this, we first needed to discover that the version of kibana installed on all nodes handling web requests for kibana:5601 were not running the same version.

sandstrom commented:
If you would prefer a method for loading dashboards and indexpatterns from the file-system (for use with Chef, Puppet, Salt, etc) please vouch for it here: https://github.com/elastic/kibana/issues/2310
